signer/src/storage/postgres.rs

```diff
@@ -1749,6 +1750,88 @@ impl super::DbRead for PgStore {
 
         self.get_sweep_transaction_by_txid(&txid).await
     }
+
+    async fn get_latest_unconfirmed_sweep_transactions(
+        &self,
+        prevout_txid: &model::BitcoinTxId,
+    ) -> Result<Vec<model::SweepTransaction>, Error> {
+        // Attempt to find the first sweep transaction in the chain which has
+        // the provided `prevout_txid` as its input. Since there can be multiple
+        // sweep transactions for the same prevout (if there was an RBF), we
+        // sort these by the created_at timestamp and take the latest one.
+        let first = sqlx::query_scalar::<_, model::BitcoinTxId>(
+            "
+            SELECT txid 
+            FROM sweep_transactions 
+            WHERE signer_prevout_txid = $1
+            ORDER BY created_at DESC
+            LIMIT 1;
+        ",
+        )
+        .bind(prevout_txid)
+        .fetch_optional(&self.0)
+        .await
+        .map_err(Error::SqlxQuery)?;
+
+        // If we didn't find any matching sweep transactions, we return an empty
+        // list.
+        let Some(first) = first else {
+            return Ok(Vec::new());
+        };
+
+        // Otherwise, we recursively find all sweep transaction ids that are
+        // chained from the first one. As a precaution, this query also filters
+        // out any sweep transactions that have already been confirmed on the
+        // Bitcoin blockchain.
+        let infos: Vec<(model::BitcoinTxId, i32)> = sqlx::query_as(
+            "
+            WITH RECURSIVE sweep_txs AS (
+                SELECT
+                    txid
+                  , signer_prevout_txid
+                  , 1 AS number
+                FROM sweep_transactions
+                WHERE 
+                    txid = $1
+
+                UNION ALL
+
+                SELECT
+                    tx.txid
+                  , tx.signer_prevout_txid
+                  ,  last.number + 1
+                FROM sweep_transactions tx
+                INNER JOIN sweep_txs last 
+                    ON tx.signer_prevout_txid = last.txid
+            )
+            SELECT
+                sweep_txs.txid
+              , number
+            FROM sweep_txs
+            LEFT JOIN bitcoin_transactions AS btc_tx 
+                ON btc_tx.txid = sweep_txs.txid
```

Comment:
What happens in case of forks here? eg a sweep tx is confirmed on a block, but then we get two block on another fork (making this second one the canonical one): the sweep tx may still be in the mempool for this second (canonical) fork, but here we would exclude it.

Reply:
Ahhh, good point... we need to join on the canonical bitcoin blockchain... nuts.

Reply:
Welp... I guess sooner or later someone was going to have to write a canonical function for transactions... :tada: